### ISSUE TITLE
Feature Request: Coupon statistics (only coupon/user/batch/time stats in this request)

### ISSUE THREAD

**Issue by imouzon** [imouzon/dmc2015#4]
Ideally, in this request I am only looking for features that can be generated using the three coupon columns, the batch, the time columns (order, received, etc), and knowing whether or not the column was used. 

Feel free to add more, either in this request or if you take up this request in your personal results.

You may find the datasets melted_train_simple_name.csv and melted_test_simple_name.csv helpful in gathering statistics in this case. I have tried to indicate if training/test set will be an issue. In almost all cases, it is probably best to calculate the statistics including all observations across datasets.
# For each coupon (regardless of column)
## Single coupon
- nCoupon = the number of times the coupon is used total (across data sets)
- propCoupon = what proportion of coupons are this type of coupon
- nCouponCol1 = how many times was this coupon seen in column 1
- nCouponCol2 = how many time was this coupon seen in column 2
- nCouponCol3 = how many times was this coupon seen in column 3
- propCol1 = what proportion of times was this coupon seen in column 1
- propCol2 = what proportion of times was this coupon seen in column 2
- propCol3 = what proportion of times was this coupon seen in column 3

*\* additionally*\* it would be nice to come up with some metric for a coupons rarity, or the rarity of it appearing in a certain column. Something like 
- propCouponLikeThis = the proportion of coupons that are like this coupon
  but I'm not 100% that is the way to do it.
## Single coupon knowing it's usage
- propCouponUsed = the proportion of times a coupon is used in the training set alone
## Single coupon and it's time
- firstTimeCouponRec = first time (date and time) the coupon we received in the dataset
- firstTimeCouponUsed = first order time (data and time) a coupon was used in the dataset
- nCouponBatch = how many times was this coupon seen in this batch
  **and so on**
  For instance, last time a coupon gets used could be interesting, but how do you calculate it?
## coupon-coupon interactions
- nCouponTrio = number of times the same values of couponID1 couponID2 and couponID3 occur
- nCoupon12pair = number of times the same values of couponID1 and couponID2 occur
- nCoupon13pair = ...
- nCoupon23pair = ...
- propCoupon12pairUsed1 = when these coupons are paired IN THE TRAINING SET, what proportion of times is coupon 1 used
- propCoupon12pairUsed2 = when these coupons are paired IN THE TRAINING SET, what proportion of times is coupon 2 used
- propCoupon12pairUsedAll = when these coupons are paired IN THE TRAINING SET, what proportion of times are coupon 1 and coupon 2 both used
- propCoupon23pairUsed2 = when these coupons are paired IN THE TRAINING SET, what proportion of times is coupon 2 used
  **and so on**
## coupon and specific user history

nCouponUser = how many times did this user get this coupon
nCouponUserBatch = how many times did user get this coupon in this batch
nCouponUserBefore = how many times did user get this coupon before
nCouponUserUsedBefore = how many times did user use this coupon before
propCouponUserUsedBefore = proportion of times a user used this coupon before
## Coupon and General User History
- CouponUsedTwice = 1 if this coupon was ever used twice by the same user 0 if no user ever used the coupon twice
- maxCouponUsedUser = the most times a coupon was used by the same user
